### PR TITLE
Overhaul documentation

### LINKED ISSUE #312: Documentation: Document the additional form field types
This bundle includes an additional form type `document` to be used in place of `entity` that isn't documented anywhere from what I can tell. The only way I found it was from reviewing code, stackoverflow, and what seems like common knowledge for people at this point but that doesn't help the new people.

There may be other types I'm unaware of and I believe it includes additional data transformers too.

Where can I help?

_[edited for clarity, think there's only the one additional type among other things]_

### LINKED ISSUE #526: Do not mention legacy driver
https://symfony.com/doc/master/bundles/DoctrineMongoDBBundle/index.html#using-the-abstraction-layer links and uses legacy driver

### PR DESCRIPTION
**Pull request by malarzm**
Currently the bundle's documentation is long and hard to navigate. This PR aims to split it and bring it up to date with Symfony 3.4+ and its standards

- [x] Split documentation into chapters available through table of contents
- [ ] Remove bundles from examples
- [ ] Use FQCNs as service identifiers
- [ ] Rewrite chapter about MongoDB library (fixes #526)
- [ ] Verify things we talk about are still relevant (i.e. reusing connection for session storage)
- [x] Write about form's `DocumentType` (fixes #312)
- [ ] Any way to see if it works after we finish the rewrite?

I'll gladly accept help with these :)

### CODE REVIEW

File: Resources/doc/index.rst [hunk 1]
```diff
@@ -568,279 +43,13 @@ and *Tree*.
...
-* ``date``
-* ``file``
-* ``float``
-* ``int``
-* ``string``
-* ``timestamp``
-
-For more information, see Doctrine's `Mapping Types documentation`_.
-
-.. index::
```
Comment (malarzm) [inline Resources/doc/index.rst]: No clue what this does, removed :)

Comment (alcaeus) [reply Resources/doc/index.rst]: Pinging @javiereguiluz - maybe they know?

Comment (javiereguiluz) [reply Resources/doc/index.rst]: This information is used to generate the "analytical index". Here you can see it: https://symfony.com/doc/current/genindex.html  We don't publicize it because we're thinking of removing it and future doc parsers may not support it. It's very useful to create indexes for paper books, where you can display the page where some concept is explained.

Comment (alcaeus) [reply Resources/doc/index.rst]: Thanks for the quick explanation! Would you recommend keeping it or is it ok if we drop this?

**Comment by malarzm** [comment]
@alcaeus @jmikola on second thought I think that a complete overhaul will not be good for you to review. For example right now I'm tweaking code examples and that'll be impossible to spot basing on the diff only. What do you think about first merging the split and `DocumentType` note that is currently done and patching things up later?

**Comment by alcaeus** [comment]
Sounds good to me 👍

File: Resources/doc/events.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
+Registering Event Listeners and Subscribers
+===========================================
+
+Doctrine allows you to register listeners and subscribers that are notified
+when different events occur inside Doctrine's ODM. For more information,
+see Doctrine's `Event Documentation`_.
+
+.. tip::
+
+    In addition to ODM events, you may also listen on lower-level MongoDB
```
Comment (alcaeus) [inline Resources/doc/events.rst]: This no longer holds true for 2.0.

File: Resources/doc/events.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+*   **event subscriber**: Use the ``doctrine_mongodb.odm.event_subscriber`` tag
+    to register a subscriber. Subscribers are responsible for implementing
+    ``Doctrine\Common\EventSubscriber`` and a method for returning the events
+    they will observe. For this reason, this tag has no ``event`` attribute;
+    however, the ``connection``, ``priority`` and ``lazy`` attributes are
+    available.
+
+.. note::
+
+    Unlike Symfony2 event listeners, Doctrine's event manager expects each
```
Comment (alcaeus) [inline Resources/doc/events.rst]: Change `Symfony2` to `Symfony`.

File: Resources/doc/first_steps.rst [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+documents to and from MongoDB.
+
+.. sidebar:: Code along with the example
+
+    If you want to follow along with the example in this chapter, create
+    an ``AcmeStoreBundle`` via:
+
+    .. code-block:: bash
+
+        php bin/console generate:bundle --namespace=Acme/StoreBundle
```
Comment (alcaeus) [inline Resources/doc/first_steps.rst]: We should probably rewrite this entire chapter to be bundle-less.

File: Resources/doc/first_steps.rst [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    use Symfony\Component\HttpFoundation\Response;
+    // ...
+
+    public function createAction()
+    {
+        $product = new Product();
+        $product->setName('A Foo Bar');
+        $product->setPrice('19.99');
+
+        $dm = $this->get('doctrine_mongodb')->getManager();
```
Comment (alcaeus) [inline Resources/doc/first_steps.rst]: This should be changed to dependency injection.

File: Resources/doc/first_steps.rst [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+Fetching an object back out of MongoDB is even easier. For example, suppose
+you've configured a route to display a specific ``Product`` based on its
+``id`` value:
+
+.. code-block:: php
+
+    public function showAction($id)
+    {
+        $product = $this->get('doctrine_mongodb')
+            ->getRepository('AcmeStoreBundle:Product')
```
Comment (alcaeus) [inline Resources/doc/first_steps.rst]: We should switch to using `::class` constants instead of aliases.

File: Resources/doc/first_steps.rst [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+to easily fetch objects based on multiple conditions:
+
+.. code-block:: php
+
+    // query for one product matching be name and price
+    $product = $repository->findOneBy(array('name' => 'foo', 'price' => 19.99));
+
+    // query for all products matching the name, ordered by price
+    $product = $repository->findBy(
+        array('name' => 'foo'),
```
Comment (alcaeus) [inline Resources/doc/first_steps.rst]: We can use short array syntax.

File: Resources/doc/first_steps.rst [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+
+    .. code-block:: php-annotations
+
+        // src/Acme/StoreBundle/Document/Product.php
+        namespace Acme\StoreBundle\Document;
+
+        use Doctrine\ODM\MongoDB\Mapping\Annotations as MongoDB;
+
+        /**
+         * @MongoDB\Document(repositoryClass="Acme\StoreBundle\Repository\ProductRepository")
```
Comment (alcaeus) [inline Resources/doc/first_steps.rst]: We should use the `::class` constant instead of strings with class names.

File: Resources/doc/installtion.rst [hunk 9]
```diff
@@ -0,0 +1,82 @@
+Installation
+============
+
+This chapter assumes you have Composer installed globally, as explained
+in the `installation chapter`_ of the Composer documentation.
+
+.. note::
+
+   The ODM requires the PHP 7 MongoDB driver (`mongodb`).
```
Comment (alcaeus) [inline Resources/doc/installtion.rst]: Calling it the "PHP 7 MongoDB driver" is wrong: it also supports PHP 5. It's ODM itself that requires PHP 7.

File: Resources/doc/installtion.rst [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+in the `installation chapter`_ of the Composer documentation.
+
+.. note::
+
+   The ODM requires the PHP 7 MongoDB driver (`mongodb`).
+
+Install the bundle with Symfony Flex
+------------------------------------
+
+Flex recipe for the DoctrineMongoDBBundle is provided through a Contrib Recipe
```
Comment (alcaeus) [inline Resources/doc/installtion.rst]: `A Flex recipe"

File: Resources/doc/first_steps.rst (hunk 4 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: This is not outdated, I just messed up line separators. Will be done later

File: Resources/doc/first_steps.rst (hunk 5 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: Same as previous comment, I'll be fixing these separately

File: Resources/doc/first_steps.rst (hunk 6 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: Same as previous comment, I'll be fixing these separately

File: Resources/doc/first_steps.rst (hunk 7 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: Same as previous comment, I'll be fixing these separately

File: Resources/doc/first_steps.rst (hunk 8 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: Same as previous comment, I'll be fixing these separately

File: Resources/doc/installtion.rst (hunk 9 above)
Comment (malarzm) [reply Resources/doc/installtion.rst]: Changed the wording and added a link

File: Resources/doc/events.rst (hunk 3 above)
Comment (malarzm) [reply Resources/doc/events.rst]: Done. Also fixed a link to "tagging" part of docs

File: Resources/doc/events.rst (hunk 2 above)
Comment (malarzm) [reply Resources/doc/events.rst]: Removed

**Comment by malarzm** [comment]
@javiereguiluz do you know when docs for bundles are built? Or is there maybe a way for you to trigger the build on demand? I'd like to merge this PR but also I'm a bit afraid I'll break something. Perfect for me would be to see the built docs after the built and fixing (and maybe asking you for a rebuild) shortly after not to leave docs broken for a longer time.

**Comment by javiereguiluz** [comment]
@malarzm it's an automatic process, so I can't force a doc build for a bundle. The build is done several times per day (I don't know the exact details) so the new docs should be published soon after the merge.

File: Resources/doc/events.rst [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+.. note::
+
+    Each connection in Doctrine has its own event manager, which is shared with
+    document managers tied to that connection. Listeners and subscribers may be
+    registered with all event managers or just one (using the connection name).
+
+In Symfony, you can register a listener or subscriber by creating a service
+and then `tagging`_ it with a specific tag.
+
+*   **event listener**: Use the ``doctrine_mongodb.odm.event_listener`` tag to
```
Comment (javiereguiluz) [inline Resources/doc/events.rst]: I'm not sure about using a bullet list in this case. Keep in mind that each bullet is super long and it even includes code samples ... so it's hard to see it as a part of a list. I'd suggest to reword this as sections or normal paragraphs.

File: Resources/doc/first_steps.rst [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+
+    // find *all* products
+    $products = $repository->findAll();
+
+    // find a group of products based on an arbitrary column value
+    $products = $repository->findByPrice(19.99);
+
+.. note::
+
+    Of course, you can also issue complex queries, which you'll learn more
```
Comment (javiereguiluz) [inline Resources/doc/first_steps.rst]: In Symfony Docs we have removed words like "Of course", "simply", "obviously", etc. because they are hostile to newcomers. Maybe you want to do the same here and remove "of course".

File: Resources/doc/first_steps.rst [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+            throw $this->createNotFoundException('No product found for id '.$id);
+        }
+
+        $product->setName('New product name!');
+        $dm->flush();
+
+        return $this->redirectToRoute('homepage');
+    }
+
+Updating an object involves just three steps:
```
Comment (javiereguiluz) [inline Resources/doc/first_steps.rst]: Most of the times, "just" is another of those bad words for newcomers. If you remove it from this phrase, no meaning is lost and it's more friendly.

File: Resources/doc/installtion.rst [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+Install the bundle with Symfony Flex
+------------------------------------
+
+A Flex recipe for the DoctrineMongoDBBundle is provided through a Contrib Recipe
+therefore you need to allow its usage:
+
+.. code-block:: bash
+
+    composer config extra.symfony.allow-contrib true
```
Comment (javiereguiluz) [inline Resources/doc/installtion.rst]: Not sure if this is needed anymore because Flex now asks you if you want to install a contrib package (always, just this time, only for this vendor, etc.)

File: Resources/doc/mongodb_library.rst [hunk 15]
```diff
@@ -0,0 +1,65 @@
+Using the MongoDB Library
+=========================
+
+Doctrine provides a thin wrapper for the PHP MongoDB client class which can conveniently
+be fetched from the Symfony2 service container. A similar example like the tutorial in the
```
Comment (javiereguiluz) [inline Resources/doc/mongodb_library.rst]: `Symfony2` -> `Symfony` ?

File: Resources/doc/security_bundle.rst [hunk 16]
```diff
@@ -0,0 +1,25 @@
+SecurityBundle integration
+==========================
+
+A user provider is available for your MongoDB projects, working exactly the
+same as the `entity` provider described in `the cookbook`_:
```
Comment (javiereguiluz) [inline Resources/doc/security_bundle.rst]: "entity" should use double quotes:

```
``entity``
```

File: Resources/doc/events.rst (hunk 11 above)
Comment (malarzm) [reply Resources/doc/events.rst]: I'll change that to normal headings and text 👍

File: Resources/doc/first_steps.rst (hunk 12 above)
Comment (malarzm) [reply Resources/doc/first_steps.rst]: I'll create separate issue to go through all docs, don't want to do too many changes as they're hard to spot.

**Comment by malarzm** [comment]
@javiereguiluz thanks for the reply and review! I'll merge this now and hopefully it builds while I'm creating issues to take care of :)
